Automattic/pocket-casts-ios
# Implement parallel download and streaming of episode.

Fixes #

Implement a method to allow download and streaming of the same episode content while it's being streamed

This is a first step on integration to test the performance and issues with piggybacking of downloads at the same time that the asset is streaming. This PR will start the download of the episodes when you press play and will switch to the downloaded asset when it's available.

The missing part is to integrate the resulting file with the rest of the DownloadManager actions. So for example if you start a download of an already streamed file you can benefit from the file and if you have an auto-download going on and you start streaming the auto-download is cancelled and replaced by the streamed file, etc..

## To test

 - Start the app on a **real device**
 - Enable the FF `cachePlayingEpisode`
 - Check that you have disable Auto-Downloads features so that they not override the streaming downloads
 - Start playing any audio episode of a podcast ( video podcasts are not supported for this method)
 - Check on the console for log messages that start with `Media Export Session`
 - Check if you see messages like `Media Export Session -> Finished exporting to:` or `Media Export Session -> Finished with error:`
 - While testing this check if performance on the player is good (no slowdowns or breaks in sound)
 - Test this with different podcasts with different effects.
 - Test this putting the app to the background while streaming

## Checklist
